tests/integration_python/test_edge_cases.py

```diff
@@ -470,6 +470,362 @@ def test_pypi_url_fragment_in_project_deps(tmp_pixi_workspace: Path, pixi: Path)
+
+    src_dir = tmp_pixi_workspace / "src" / "test"
+    src_dir.mkdir(parents=True, exist_ok=True)
+    (src_dir / "__init__.py").touch()
+
+    # Should succeed with correct MD5 hash
+    verify_cli_command(
+        [pixi, "install", "-v"],
+        cwd=tmp_pixi_workspace,
+        expected_exit_code=ExitCode.SUCCESS,
+    )
+
+
+def test_pypi_external_source_incorrect_md5_hash(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test MD5 hash verification fails with incorrect hash"""
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "idna @ https://files.pythonhosted.org/packages/76/c6/c88e154df9c4e1a2a66ccf0005a88dfb2650c1dffb6f5ce603dfbd452ce3/idna-3.10-py3-none-any.whl#md5=00000000000000000000000000000000"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
+channels = ["https://prefix.dev/conda-forge"]
+"""
+    pyproject_path = tmp_pixi_workspace / "pyproject.toml"
+    pyproject_path.write_text(pyproject_content)
+
+    src_dir = tmp_pixi_workspace / "src" / "test"
+    src_dir.mkdir(parents=True, exist_ok=True)
+    (src_dir / "__init__.py").touch()
+
+    # Should fail with incorrect MD5 hash
+    verify_cli_command(
+        [pixi, "install", "-v"],
+        cwd=tmp_pixi_workspace,
+        expected_exit_code=ExitCode.FAILURE,
+        stderr_contains=["Hash verification failed"],
+    )
+
+
+def test_pypi_external_source_no_hash(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test that direct URL without hash fragment works normally"""
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "requests @ https://files.pythonhosted.org/packages/f9/9b/335f9764261e915ed497fcdeb11df5dfd6f7bf257d4a6a2a686d80da4d54/requests-2.32.3-py3-none-any.whl"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
+channels = ["https://prefix.dev/conda-forge"]
+"""
+    pyproject_path = tmp_pixi_workspace / "pyproject.toml"
+    pyproject_path.write_text(pyproject_content)
+
+    src_dir = tmp_pixi_workspace / "src" / "test"
+    src_dir.mkdir(parents=True, exist_ok=True)
+    (src_dir / "__init__.py").touch()
+
+    # Should succeed without hash
+    verify_cli_command(
+        [pixi, "install", "-v"],
+        cwd=tmp_pixi_workspace,
+        expected_exit_code=ExitCode.SUCCESS,
+    )
+
+
+def test_pypi_external_source_invalid_hash_format(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test that invalid hash format in URL fragment is handled gracefully"""
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "requests @ https://files.pythonhosted.org/packages/f9/9b/335f9764261e915ed497fcdeb11df5dfd6f7bf257d4a6a2a686d80da4d54/requests-2.32.3-py3-none-any.whl#sha256=notahexhash"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
```

Comment:
To reduce the overhead of testing we could include only the current platform here.